provider/openstack/provider.go

```diff
@@ -1039,17 +1039,58 @@ func (e *Environ) StartInstance(args environs.StartInstanceParams) (*environs.St
 		args.InstanceConfig.MachineId,
 	)
 
+	waitForActiveServerDetails := func(
+		client *nova.Client,
+		id string,
+	) (server *nova.ServerDetail, err error) {
+		logger.Debugf("waiting for instance %s to start...", id)
+
+		for {
+			server, err = client.GetServer(id)
+			if server.Status != nova.StatusBuild {
+				break
+			}
+			// We don't want to flood the connection while polling the server waiting
+			// for it to start.
+			logger.Debugf("server has status %s, waiting 10 seconds before polling again...", server.Status)
+			time.Sleep(10 * time.Second)
+		}
+		return server, err
+	}
+
 	tryStartNovaInstance := func(
 		attempts utils.AttemptStrategy,
 		client *nova.Client,
 		instanceOpts nova.RunServerOpts,
 	) (server *nova.Entity, err error) {
+		var message string
 		for a := attempts.Start(); a.Next(); {
 			server, err = client.RunServer(instanceOpts)
-			if err == nil || gooseerrors.IsNotFound(err) == false {
+			if err != nil {
```

Comment:
No, if the error IsNotFound we still break out of the loop, as we did previously.  Now, any error from RunServer causes the provider to drop out of tryStartNovaInstance().